File: CS315-Programming Languages/HW2/Codes/22101782_akin_yasemin.go
package main

import "fmt"

func outerFunction() func() {
	outerVar := 10
	packageLevelVar := &globalVar

	closure := func() {
		fmt.Println("Outer variable:", outerVar)

		fmt.Println("Package-level variable:", *packageLevelVar)
	}

	return closure
}

var globalVar = 100

func outerClosure() func() func() string {
	outerVar := "Outer"

	return func() func() string {
		innerVar := "Inner"
		fmt.Println(outerVar) // Accesses outer closure variable

		return func() string {
			return outerVar + " " + innerVar // Accesses both outer and inner variables
		}
	}
}

func main() {

	//1
	adderFunc := func(x int) func(int) int {
		return func(y int) int {
			return x + y
		}
	}

	addFive := adderFunc(5)
	fmt.Println(addFive(10)) // Output: 15

	//2
	myClosure := outerFunction()
	myClosure() // Access the outer and package-level variables

	//3
	outerVar := 10
	closure := func() {
		outerVar += 5 // Modifying the outer variable
		fmt.Println("Inside Closure:", outerVar)
	}

	fmt.Println("Before Closure:", outerVar)
	closure()
	fmt.Println("After Closure:", outerVar)

	//4
	innerClosure := outerClosure()()
	fmt.Println(innerClosure()) // Executes the nested closure
}

File: CS315-Programming Languages/HW2/Codes/22101782_akin_yasemin.dart
String Function(int) createAdder(int addBy) {
    return (int i) => 'Result: ${i + addBy}';
}

String outerFunction(String message) {
  // An outer scope variable
  String outerVariable = 'Outer: $message';

  // Closure definition
  String innerFunction() {
    // Accessing the outer scope variable
    return 'Inner accesses: $outerVariable';
  }

  return innerFunction(); // Executing the closure
}

var globalVar = 10;

Function outerClosure() {
  var outerVar = 'Outer';

  Function innerClosure() {
    var innerVar = 'Inner';
    print(outerVar); // Accesses variable from the outer closure
    return () {
      print('$outerVar $innerVar'); // Accesses variables from both scopes
    };
  }

  return innerClosure();
}

void main() {

  //1
  var addTwo = createAdder(2);
  print(addTwo(3));  // Output: Result: 5
  
  //2
  var result = outerFunction('Hello');
  print(result); // Output will be displayed
  
  //3
  final outerFinal = 10;
  var outerVar = 10;
    Function closure = () {
        // outerFinal = 20; // This line would cause an error, as final variables can't be modified.
        print(outerFinal);
        outerVar = 20;
        print(outerVar);
        globalVar = 20;
        print(globalVar);
    };
    closure();  // Outputs: 10 20 20
  
  //4
  var nestedClosure = outerClosure();
  nestedClosure(); // Executes the nested closure
}